router/internal/eventorder/eventorder.go

```diff
@@ -0,0 +1,255 @@
+		if barrier.failedJobID != nil {
+			failedJobID = fmt.Sprintf("%d", *barrier.failedJobID)
+		}
+		sb.WriteString(fmt.Sprintf("{userID: %s, failedJobID: %v, concurrentJobs: %v}", userID, failedJobID, barrier.concurrencyLimiter))
+	}
+	sb.WriteString("]}")
+	return sb.String()
+}
+
+type barrierInfo struct {
+	failedJobID        *int64             // nil if no failed job
+	mu                 sync.RWMutex       // protects concurrentJobs
+	concurrencyLimiter map[int64]struct{} // nil if concurrency limiter is off
+}
+
+type command interface {
+	enqueue(barriers map[string]*barrierInfo) bool
+	execute(barriers map[string]*barrierInfo)
+}
+
+type cmd struct {
+	userID string
+	jobID  int64
+}
+
+// default behaviour is to try and remove the jobID from the concurrent jobs map
+func (c *cmd) execute(barriers map[string]*barrierInfo) {
+	if barrier, ok := barriers[c.userID]; ok {
+		barrier.mu.Lock()
+		defer barrier.mu.Unlock()
+		delete(barrier.concurrencyLimiter, c.jobID)
+	}
+}
+
+// default behaviour is to enqueue the command if a barrier for this userID already exists
+func (c *cmd) enqueue(barriers map[string]*barrierInfo) bool {
+	_, ok := barriers[c.userID]
+	return ok
+}
+
+// jobFailedCommand is a command that is executed when a job has failed.
+type jobFailedCommand struct {
+	*cmd
+}
+
+// If no failed jobID is in the barrier make this jobID the failed job for this userID. Removes the job from the concurrent jobs map if it exists there
+func (c *jobFailedCommand) execute(barriers map[string]*barrierInfo) {
+	barrier, ok := barriers[c.userID]
+	if !ok {
+		barrier = &barrierInfo{}
+		barriers[c.userID] = barrier
+	}
+	barrier.mu.Lock()
+	defer barrier.mu.Unlock()
+
+	if barrier.failedJobID == nil {
+		barrier.failedJobID = &c.jobID
+	} else if *barrier.failedJobID > c.jobID {
+		panic(fmt.Errorf("previousFailedJob:%d > jobID:%d", *barrier.failedJobID, c.jobID))
+	}
+	// reset concurrency limiter
+	barrier.concurrencyLimiter = nil
+}
+
+// a failed command never gets enqueued
+func (*jobFailedCommand) enqueue(_ map[string]*barrierInfo) bool {
+	return false
+}
+
+// jobSucceededCmd is a command that is executed when a job has succeeded.
+type jobSucceededCmd struct {
+	*cmd
+}
+
+// removes the barrier for this userID, if it exists
+func (c *jobSucceededCmd) execute(barriers map[string]*barrierInfo) {
+	if barrier, ok := barriers[c.userID]; ok {
+		if barrier.failedJobID != nil && *barrier.failedJobID != c.jobID { // out-of-sync command (failed commands get executed immediately)
+			return
+		}
+	}
+	delete(barriers, c.userID)
+}
+
+// JobAbortedCommand is a command that is executed when a job has aborted.
+type JobAbortedCommand struct {
+	*cmd
+}
+
+// Creates a concurrent jobs map if none exists. Also removes the jobID from the concurrent jobs map if it exists there
+func (c *JobAbortedCommand) execute(barriers map[string]*barrierInfo) {
+	if barrier, ok := barriers[c.userID]; ok {
+		if barrier.failedJobID == nil {
```

Comment:
Why do we not need a read lock over here?

Reply:
We have already acquired a full `Barrier` lock in the caller, see `Barrier#StateChanged` and `Barrier#Sync`.
In case you are referring to the other lock, `barrierInfo.mu`, this lock is only there to protect `barrierInfo.concurrencyLimiter` as mentioned in its definition comment